service/storage_service.cc

```diff
@@ -459,15 +459,15 @@ future<storage_service::nodes_to_notify_after_sync> storage_service::sync_raft_t
         auto ip = am.find(id);
 
         rtlogger.trace("loading topology: raft id={} ip={} node state={} dc={} rack={} tokens state={} tokens={} shards={}",
-                      id, ip, rs.state, rs.datacenter, rs.rack, _topology_state_machine._topology.tstate, rs.ring.value().tokens, rs.shard_count, rs.cleanup);
```

Comment:
Some of those `values()` that you remove was used specifically to gracefully catch cases where the tokens should be present but they do not. This change makes the code less safe. regarding the claim:

>  The reason is that tokens are stored in a non-frozen set, which does not differentiate an empty set from null

Is this really the case? Is deleting (or never assigning) a value to a cell will be seen as an empty set?

Reply:
> > The reason is that tokens are stored in a non-frozen set, which does not differentiate an empty set from null
> 
> Is this really the case? Is deleting (or never assigning) a value to a cell will be seen as an empty set?

https://opensource.docs.scylladb.com/stable/cql/types.html#sets

*Note that ScyllaDB does not distinguish an empty set from a missing value, thus assigning an empty set ({}) to a set is the same as deleting it.*

I also checked it manually. This statement is true only for non-frozen sets.

> Some of those `values()` that you remove was used specifically to gracefully catch cases where the tokens should be present but they do not. This change makes the code less safe. regarding the claim:

I'm not happy about it, but I don't think there is a simple way to avoid it. A normal node is zero-token if it has an empty set of tokens. Any other condition would require extending the topology (for example, by some boolean value) and treating zero-token nodes differently everywhere, so we don't want to do it. Since we look at tokens and an empty set of tokens is indistinguishable from null, we must assume that if a normal node has zero-tokens, then it is a correct zero-token node.

Reply:
> > > The reason is that tokens are stored in a non-frozen set, which does not differentiate an empty set from null
> > 
> > 
> > Is this really the case? Is deleting (or never assigning) a value to a cell will be seen as an empty set?
> 
> https://opensource.docs.scylladb.com/stable/cql/types.html#sets
> 
> _Note that ScyllaDB does not distinguish an empty set from a missing value, thus assigning an empty set ({}) to a set is the same as deleting it._
> 
> I also checked it manually. This statement is true only for non-frozen sets.

So if I assign empty set to a `token` cell `row.has("tokens")` will return false? This is weird.

> 
> > Some of those `values()` that you remove was used specifically to gracefully catch cases where the tokens should be present but they do not. This change makes the code less safe. regarding the claim:
> 
> I'm not happy about it, but I don't think there is a simple way to avoid it. A normal node is zero-token if it has an empty set of tokens. Any other condition would require extending the topology (for example, by some boolean value) and treating zero-token nodes differently everywhere, so we don't want to do it. Since we look at tokens and an empty set of tokens is indistinguishable from null, we must assume that if a normal node has zero-tokens, then it is a correct zero-token node.

We can add bool and leave optional<ring> as it is. The optional will be set to contain an empty set if the bool is true when the ring value is assigned. Would it simplify the code I am not sure, but it will make it possible to distinguish between cases where the token value is not assigned yet and cases where it suppose to be assigned but empty because the node is zero token.

Reply:
> So if I assign empty set to a token cell row.has("tokens") will return false? This is weird.

That's how non-frozen collections work in Scylla/Cassandra. Empty set = 0 cells = nothing.

> We can add bool and leave optional as it is. The optional will be set to contain an empty set if the bool is true when the ring value is assigned. Would it simplify the code I am not sure, but it will make it possible to distinguish between cases where the token value is not assigned yet and cases where it suppose to be assigned but empty because the node is zero token.

Adding a bool makes more incorrect states representable in code. I like the approach taken here: normal node with zero tokens == it's a zero-token node, normal node with non-zero tokens == it's a token-owning node. Looks pretty straightforward.

For distinguishing between the cases, it's already possible by looking at node state (is it none/bootstrapping/normal etc.)

Reply:
> > So if I assign empty set to a token cell row.has("tokens") will return false? This is weird.
> 
> That's how non-frozen collections work in Scylla/Cassandra. Empty set = 0 cells = nothing.
> 

May be we should have used something more sane then :(

> > We can add bool and leave optional as it is. The optional will be set to contain an empty set if the bool is true when the ring value is assigned. Would it simplify the code I am not sure, but it will make it possible to distinguish between cases where the token value is not assigned yet and cases where it suppose to be assigned but empty because the node is zero token.
> 
> Adding a bool makes more incorrect states representable in code. 

I am not sure this is the case. The only possible additional incorrect state would be `zero_token=true` but `!ring.value(),empty()`. By making ring non optional we made a lot of incorrect states undetectable.

> I like the approach taken here: normal node with zero tokens == it's a zero-token node, normal node with non-zero tokens == it's a token-owning node. Looks pretty straightforward.
> 
> For distinguishing between the cases, it's already possible by looking at node state (is it none/bootstrapping/normal etc.)

We already look at node state, but we are also able to detect that a node state does not much the ring state gracefully. We drop this ability here. This is fine assuming we never add bugs. But we usually do.